File: ios/MegaPDF/DemoContent.swift
import CoreGraphics
import Foundation
import UIKit

/// Marketing/screenshot support (`-screenshot <state>` launch argument):
/// seeds believable content so App Store captures show the real app doing
/// real work. Never active in normal launches.
enum DemoContent {

    static var requestedState: String? {
        let args = ProcessInfo.processInfo.arguments
        guard let i = args.firstIndex(of: "-screenshot"), i + 1 < args.count else { return nil }
        return args[i + 1]
    }

    /// Everything the screenshots show follows the app language through the
    /// String Catalog (#91): a `-AppleLanguages (fr-CA)` launch opens the French
    /// agreement (`demo-fr.pdf`) under French names, and searches for a French
    /// word. The keys are the English values.

    /// The bundled demo agreement's resource name, without extension.
    static var demoResource: String { String(localized: "demo", comment: "screenshot demo PDF resource name") }

    /// The same agreement with nothing filled in (`demo-blank.pdf`,
    /// `demo-fr-blank.pdf`): what `-screenshot story` opens, so the preview
    /// video can tick, sign and type on camera instead of over a finished page.
    static var blankDemoResource: String { demoResource + "-blank" }

    /// The demo agreement's display name in the title bar and the recents list.
    static var documentName: String { String(localized: "Rental Agreement.pdf") }

    /// Term seeded into the find bar by `-screenshot search`. "rental" is the
    /// most-repeated word on the demo agreement's only page — it hits the
    /// "Equipment Rental Agreement" heading, "Sunrise Tool Rental" and "the
    /// rental equipment" — so the capture shows three highlights clustered
    /// under the bar with the counter reading "1 of 3". Search is
    /// case-insensitive, so the lower-case term matches the capitalized ones.
    /// The French page repeats "location" the same three times.
    static var searchTerm: String { String(localized: "rental", comment: "screenshot search term; must occur three times on the demo page") }

    /// What the Redact capture marks (#173): a line of the demo agreement with something on
    /// it worth removing. Marked by what it says rather than by a rectangle, so the shot
    /// lands on a sentence in every language.
    static var redactedWord: String {
        String(localized: "customer named",
               comment: "#173: what the Redact capture marks; a phrase on the demo agreement's first line")
    }

    /// Marketing "Add text" shot (#43): the name the customer would print under
    /// the signature rule the demo agreement draws at y=400, and where it sits.
    ///
    /// Localised like everything else here (#146 §3). Dave, 2026-09-15: the
    /// French captures must not show "Jane Whitfield", and the French names
    /// carry accents — Hélène Bélanger for fr-CA, Céline Lefèvre for fr — which
    /// is also what proves the accented glyphs survive the face the demo picks.
    /// `MegaPDFUITests/DemoFlowUITests.swift` varies its copy the same way.
    static var printedName: String {
        String(localized: "Jane Whitfield", comment: "screenshot demo person; French captures use French names, with accents (#146 §3)")
    }
    static let printedNameX: Double = 72
    static let printedNameY: Double = 372

    /// The handwritten "MegaWoman" demo signature (bundled transparent PNG,
    /// rendered from a script typeface at build time).
    static func signatureImage() -> CGImage? {
        guard let url = Bundle.main.url(forResource: "demo-signature", withExtension: "png"),
              let data = try? Data(contentsOf: url) else { return nil }
        return UIImage(data: data)?.cgImage
    }

    /// Somewhere believable for a demo document to live (#165). Localised like the
    /// rest of the demo content: a French capture reads "Téléchargements", because
    /// that is what the Files app would have called the folder.
    static var iCloudDrive: String { String(localized: "iCloud Drive") }
    static var downloadsFolder: String {
        String(localized: "Downloads", comment: "#165: demo recents; the Files app's Downloads folder")
    }
    static var clientFolder: String {
        String(localized: "Clients", comment: "#165: demo recents; a folder of client paperwork")
    }
    static var archiveFolder: String {
        String(localized: "Archive", comment: "#165: demo recents; a folder of older paperwork")
    }

    static func demoRecents() -> [RecentEntry] {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let day: Int64 = 86_400_000
        return [
            (documentName, now - day / 2, [iCloudDrive, clientFolder]),
            (String(localized: "Field Trip Permission.pdf"), now - 2 * day,
             [RecentLocation.deviceName(), downloadsFolder]),
            (String(localized: "Insurance Claim Form.pdf"), now - 6 * day, [iCloudDrive]),
        ].map { name, at, segments in
            RecentEntry(bookmarkBase64: Data(name.utf8).base64EncodedString(),
                        displayName: name, lastOpenedEpochMs: at,
                        location: RecentLocation(segments: segments))
        }
    }

    /// `-screenshot recents`: one file name from four places, one of them gone.
    ///
    /// The whole of #165 on a single screen — four rows nobody could have told apart
    /// before, and the greyed-out one the Files app would show for a file that has
    /// moved. Made up rather than real, because no simulator has an iCloud account or
    /// a second file provider, and because a capture has to look the same every time
    /// it is taken.
    static func recentsScenario() -> (entries: [RecentEntry], unavailable: Set<String>) {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let day: Int64 = 86_400_000
        // Surnames are not translated; the folders around them are.
        let places: [(segments: [String], at: Int64)] = [
            ([iCloudDrive, "Smith"], now - day / 4),
            ([iCloudDrive, "Jones"], now - day),
            ([RecentLocation.deviceName(), downloadsFolder], now - 3 * day),
            ([iCloudDrive, archiveFolder], now - 9 * day),
        ]
        let entries = places.enumerated().map { index, place in
            RecentEntry(bookmarkBase64: Data("recents-\(index)".utf8).base64EncodedString(),
                        displayName: documentName,
                        lastOpenedEpochMs: place.at,
                        location: RecentLocation(segments: place.segments))
        }
        // The last one has moved since it was opened.
        return (entries, Set(entries.suffix(1).map(\.id)))
    }
}

File: src/MegaPDF.Avalonia/DemoContent.cs
using System.Globalization;

namespace MegaPDF.Avalonia;

/// <summary>
/// The content the store captures put *into* the demo document — as opposed to
/// the app's own UI text, which lives in Strings.resx.
///
/// It is deliberately not in the string catalogue: nobody using MegaPDF ever
/// sees any of this. It only exists so `--story` and `--screenshot-state
/// textbox` can pose a French capture that is French all the way through,
/// rather than a French window over an English name (#146 §3).
///
/// The same values, chosen by the same rule, are in `ios/MegaPDF/DemoContent.swift`
/// and `tools/macos-record-demo.sh` — keep the three in step.
/// </summary>
internal static class DemoContent
{
    /// <summary>
    /// Who signs the demo agreement. Dave, 2026-09-15: the French captures must
    /// not show "Jane Whitfield", and the names must carry accents — which is
    /// also what proves the accented glyphs survive whatever face the demo picks.
    /// </summary>
    internal static string PrintedName => ForLanguage(
        english: "Jane Whitfield", frenchCanadian: "Hélène Bélanger", french: "Céline Lefèvre");

    /// <summary>
    /// What the find steps search for. Three hits on the demo agreement in both
    /// languages, so "1 of 3" reads the same — see tools/gen_test_fixtures.py,
    /// where the French page is translated to keep exactly that.
    /// </summary>
    internal static string SearchTerm => ForLanguage(
        english: "rental", frenchCanadian: "location", french: "location");

    /// <summary>
    /// What the Redact capture marks (#173): a line of the demo agreement with something on
    /// it worth removing. Marked by what it says rather than by a rectangle, so the shot
    /// lands on a sentence in every language rather than on whatever a fixed box covers.
    /// </summary>
    internal static string RedactedWord => ForLanguage(
        english: "customer named", frenchCanadian: "client nommé", french: "client nommé");

    /// <summary>
    /// What the demo document is called in a capture.
    ///
    /// The fixture on disk is demo.pdf / demo-fr.pdf, and that name is in the status
    /// line of every shot that has a document open — beside a home shot whose recents
    /// said "Contrat de location.pdf". One set, two names for the same file (#146 §3).
    /// The capture script copies the fixture to this name before it opens it.
    /// </summary>
    internal static string DocumentFileName => ForLanguage(
        english: "Rental Agreement.pdf",
        frenchCanadian: "Contrat de location.pdf",
        french: "Contrat de location.pdf");

    /// <summary>
    /// The rows the home screenshot shows (#146 §3): a name and the place under it.
    ///
    /// Made up rather than read from the machine, for the reason iOS's demoRecents
    /// gives: a capture has to look the same every time it is taken, and the machine's
    /// own list is whatever it last opened — which on the capture Mac was a path
    /// through the sandbox container. The first row is the document the rest of the
    /// set is about; the other two are there so the list is a list.
    ///
    /// The place lines use the same separator the real rows do, so nothing about the
    /// shot is a special case in the view.
    /// </summary>
    internal static IReadOnlyList<(string Name, string Location)> Recents =>
    [
        (DocumentFileName,
         ForLanguage("Documents › Clients", "Documents › Clients", "Documents › Clients")),
        (ForLanguage("Field Trip Permission.pdf", "Autorisation de sortie scolaire.pdf", "Autorisation de sortie scolaire.pdf"),
         ForLanguage("Downloads", "Téléchargements", "Téléchargements")),
        (ForLanguage("Insurance Claim Form.pdf", "Formulaire de réclamation.pdf", "Formulaire de réclamation.pdf"),
         // The one row whose place is not the same on both of this project's desktops.
         // "iCloud Drive" is where a Mac keeps a file and is right in the Mac listing's
         // home shot; on a Linux listing it advertises an Apple service the app cannot
         // reach, in a screenshot whose whole job is to show the app on this platform
         // (#254 A1). The macOS value is untouched, so the signed-off Mac set is
         // unchanged byte for byte.
         OperatingSystem.IsLinux()
             ? ForLanguage("Desktop", "Bureau", "Bureau")
             : ForLanguage("iCloud Drive", "iCloud Drive", "iCloud Drive")),
    ];

    private static string ForLanguage(string english, string frenchCanadian, string french)
    {
        var culture = CultureInfo.CurrentUICulture;
        if (!culture.TwoLetterISOLanguageName.Equals("fr", StringComparison.OrdinalIgnoreCase))
            return english;
        // fr-CA is its own listing language, with its own name; every other
        // French locale takes the France one, as the catalogue does.
        return culture.Name.Equals("fr-CA", StringComparison.OrdinalIgnoreCase) ? frenchCanadian : french;
    }
}
